processing/src/main/java/org/apache/druid/segment/join/HashJoinEngine.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import org.apache.druid.query.BaseQuery;
+import org.apache.druid.query.dimension.DimensionSpec;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.Cursor;
+import org.apache.druid.segment.DimensionSelector;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.joda.time.DateTime;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class HashJoinEngine
+{
+  private HashJoinEngine()
+  {
+    // No instantiation.
+  }
+
+  /**
+   * Creates a cursor that represents the join of {@param leftCursor} with {@param joinableClause}. The resulting
+   * cursor may generate nulls on the left-hand side (for righty joins; see {@link JoinType#isRighty()}) or on
+   * the right-hand side (for lefty joins; see {@link JoinType#isLefty()}). Columns that start with the
+   * joinable clause's prefix (see {@link JoinableClause#getPrefix()}) will come from the Joinable's column selector
+   * factory, and all other columns will come from the leftCursor's column selector factory.
+   *
+   * Ensuing that the joinable clause's prefix does not conflict with any columns from "leftCursor" is the
```

Comment:
`Ensuring`?

Reply:
Would you elaborate more on this, especially what conflict between the prefix of the joinable clause and columns from the left cursor? Is it like the conflict between the right table name and the left column name?

Reply:
Oops, yeah, that's a typo. It should be "ensuring".

Is this clearer?

```java
   * Ensuring that the joinable clause's prefix does not conflict with any columns from "leftCursor" is the
   * responsibility of the caller. If there is such a conflict (for example, if the joinable clause's prefix is "j.",
   * and the leftCursor has a field named "j.j.abrams"), then the field from the leftCursor will be shadowed and will
   * not be queryable through the returned Cursor. This happens even if the right-hand joinable doesn't actually have a
   * column with this name.
```